components/esp_eth/src/esp_eth_phy_ksz8041.c

```diff
@@ -0,0 +1,354 @@
+    return ESP_OK;
+err:
+    return ESP_FAIL;
+}
+
+static esp_err_t ksz8041_pwrctl(esp_eth_phy_t *phy, bool enable)
+{
+    phy_ksz8041_t *ksz8041 = __containerof(phy, phy_ksz8041_t, parent);
+    esp_eth_mediator_t *eth = ksz8041->eth;
+    bmcr_reg_t bmcr;
+    PHY_CHECK(eth->phy_reg_read(eth, ksz8041->addr, ETH_PHY_BMCR_REG_ADDR, &(bmcr.val)) == ESP_OK,
+              "read BMCR failed", err);
+    if (!enable) {
+        /* General Power Down Mode */
+        bmcr.power_down = 1;
+    } else {
+        /* Normal operation Mode */
+        bmcr.power_down = 0;
+    }
+    PHY_CHECK(eth->phy_reg_write(eth, ksz8041->addr, ETH_PHY_BMCR_REG_ADDR, bmcr.val) == ESP_OK,
+              "write BMCR failed", err);
+    PHY_CHECK(eth->phy_reg_read(eth, ksz8041->addr, ETH_PHY_BMCR_REG_ADDR, &(bmcr.val)) == ESP_OK,
+              "read BMCR failed", err);
+    if (!enable) {
+        PHY_CHECK(bmcr.power_down == 1, "power down failed", err);
+    } else {
+        PHY_CHECK(bmcr.power_down == 0, "power up failed", err);
+    }
+    return ESP_OK;
+err:
+    return ESP_FAIL;
+}
+
+static esp_err_t ksz8041_set_addr(esp_eth_phy_t *phy, uint32_t addr)
+{
+    phy_ksz8041_t *ksz8041 = __containerof(phy, phy_ksz8041_t, parent);
+    ksz8041->addr = addr;
+    return ESP_OK;
+}
+
+static esp_err_t ksz8041_get_addr(esp_eth_phy_t *phy, uint32_t *addr)
+{
+    PHY_CHECK(addr, "addr can't be null", err);
+    phy_ksz8041_t *ksz8041 = __containerof(phy, phy_ksz8041_t, parent);
+    *addr = ksz8041->addr;
+    return ESP_OK;
+err:
+    return ESP_ERR_INVALID_ARG;
+}
+
+static esp_err_t ksz8041_del(esp_eth_phy_t *phy)
+{
+    phy_ksz8041_t *ksz8041 = __containerof(phy, phy_ksz8041_t, parent);
+    free(ksz8041);
+    return ESP_OK;
+}
+
+static esp_err_t ksz8041_init(esp_eth_phy_t *phy)
+{
+    phy_ksz8041_t *ksz8041 = __containerof(phy, phy_ksz8041_t, parent);
+    esp_eth_mediator_t *eth = ksz8041->eth;
+    /* Power on Ethernet PHY */
+    PHY_CHECK(ksz8041_pwrctl(phy, true) == ESP_OK, "power control failed", err);
+    /* Reset Ethernet PHY */
+    PHY_CHECK(ksz8041_reset(phy) == ESP_OK, "reset failed", err);
+    /* Check PHY ID */
+    phyidr1_reg_t id1;
+    phyidr2_reg_t id2;
+    PHY_CHECK(eth->phy_reg_read(eth, ksz8041->addr, ETH_PHY_IDR1_REG_ADDR, &(id1.val)) == ESP_OK,
+              "read ID1 failed", err);
+    PHY_CHECK(eth->phy_reg_read(eth, ksz8041->addr, ETH_PHY_IDR2_REG_ADDR, &(id2.val)) == ESP_OK,
+              "read ID2 failed", err);
+    PHY_CHECK(id1.oui_msb == 0x22 && id2.oui_lsb == 0x5 && id2.vendor_model == 0x11, "wrong chip ID", err);
+    return ESP_OK;
+err:
+    return ESP_FAIL;
+}
+
+static esp_err_t ksz8041_deinit(esp_eth_phy_t *phy)
+{
+    /* Power off Ethernet PHY */
+    PHY_CHECK(ksz8041_pwrctl(phy, false) == ESP_OK, "power control failed", err);
+    return ESP_OK;
+err:
+    return ESP_FAIL;
+}
+
+esp_eth_phy_t *esp_eth_phy_new_ksz8041(const eth_phy_config_t *config)
```

Comment:
Could please add the declaration of this API in `esp_eth_phy.h`?

Reply:
Sure